pkg/ingester/retention.go

```diff
@@ -359,6 +359,32 @@ type realFSBlockManager struct {
 	FS      fileSystem
 }
 
+func (bm *realFSBlockManager) getUploadedBlockIds(tenantID string) (map[ulid.ULID]struct{}, error) {
+	localDirPath := filepath.Join(bm.Root, tenantID, phlareDBLocalPath)
+
+	shipperPath := filepath.Join(localDirPath, shipper.MetaFilename)
+	bytes, err := fs.ReadFile(bm.FS, shipperPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return make(map[ulid.ULID]struct{}), nil
+		}
+		return nil, err
+	}
+
+	var meta shipper.Meta
+	err = json.Unmarshal(bytes, &meta)
+	if err != nil {
+		return nil, err
+	}
+
+	uploadedBlockIDs := make(map[ulid.ULID]struct{}, len(meta.Uploaded))
+	for _, id := range meta.Uploaded {
+		uploadedBlockIDs[id] = struct{}{}
+	}
+
+	return uploadedBlockIDs, nil
+}
```

Comment:
@kolesnikovae, Extracted out abstraction for getting list of uploaded blocks ids, and when `shipper metafile` does not exist, it ignores and return empty map. This mostly solves problem, as without any information of uploaded, it will work fine only the `uploaded` flag will be unset and it will not affect any flows, `CleanupBlocksWhenHighDiskUtilization` will work fine as it sorts by both age and uploaded(which will be false for given scenario)